File: Information_Aggregation/backend/app/api/v1/permissions.py
from fastapi import APIRouter, HTTPException, Query, status

from app.api.deps import AdminUser, CurrentUser, DbSession
from app.schemas import PageResult, ResponseBase
from app.schemas.user import (
    AccessReviewAction,
    SystemSettingOut,
    ViewAccessRequestCreate,
    ViewAccessRequestOut,
)
from app.services.permission_service import PermissionService

router = APIRouter(prefix="/permissions", tags=["权限管理"])


@router.post("/access-requests", response_model=ResponseBase[ViewAccessRequestOut], status_code=status.HTTP_201_CREATED)
def submit_access_request(db: DbSession, user: CurrentUser, data: ViewAccessRequestCreate):
    try:
        req = PermissionService.create_access_request(db, user, data)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    return ResponseBase(
        data=ViewAccessRequestOut(
            id=req.id,
            user_id=req.user_id,
            request_type=req.request_type,
            status=req.status,
            reason=req.reason,
            reviewer_id=req.reviewer_id,
            review_note=req.review_note,
            created_at=req.created_at,
            reviewed_at=req.reviewed_at,
            username=user.username,
            nickname=user.nickname,
        ),
        message="申请已提交",
    )


@router.get("/access-requests/types", response_model=ResponseBase[list])
def list_applicable_request_types(user: CurrentUser):
    return ResponseBase(data=PermissionService.list_request_types_for_user(user))


@router.get("/access-requests/stats", response_model=ResponseBase[dict])
def access_request_stats(db: DbSession, user: CurrentUser):
    return ResponseBase(data=PermissionService.get_access_request_stats(db, user))


@router.get("/access-requests", response_model=ResponseBase[PageResult[ViewAccessRequestOut]])
def list_access_requests(
    db: DbSession,
    user: CurrentUser,
    status: str | None = None,
    request_type: str | None = None,
    scope: str | None = Query(
        None,
        description="mine=仅我的申请；review=待我审批范围（审核员）。默认：审核员为 review，普通用户为 mine",
    ),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
):
    items, total = PermissionService.list_access_requests(
        db, user, status, request_type, page, page_size, scope=scope
    )
    return ResponseBase(
        data=PageResult(
            items=[ViewAccessRequestOut(**i) for i in items],
            total=total,
            page=page,
            page_size=page_size,
        )
    )


@router.post("/access-requests/{request_id}/review", response_model=ResponseBase[ViewAccessRequestOut])
def review_access_request(
    db: DbSession,
    reviewer: AdminUser,
    request_id: int,
    data: AccessReviewAction,
):
    try:
        req = PermissionService.review_access_request(
            db, request_id, reviewer, data.approve, data.review_note
        )
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc

    from app.models import User

    applicant = db.query(User).filter(User.id == req.user_id).first()
    reviewer = db.query(User).filter(User.id == req.reviewer_id).first() if req.reviewer_id else None
    return ResponseBase(
        data=ViewAccessRequestOut(
            id=req.id,
            user_id=req.user_id,
            request_type=req.request_type,
            status=req.status,
            reason=req.reason,
            reviewer_id=req.reviewer_id,
            review_note=req.review_note,
            created_at=req.created_at,
            reviewed_at=req.reviewed_at,
            username=applicant.username if applicant else None,
            nickname=applicant.nickname if applicant else None,
            reviewer_username=reviewer.username if reviewer else None,
            reviewer_nickname=reviewer.nickname if reviewer else None,
        ),
        message="已通过申请" if data.approve else "已拒绝申请",
    )


@router.post("/users/{user_id}/revoke-library", response_model=ResponseBase[dict])
def revoke_library_access(db: DbSession, reviewer: AdminUser, user_id: int):
    try:
        user = PermissionService.revoke_library_access(db, user_id, reviewer)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    return ResponseBase(data={"user_id": user.id, "view_library": bool(user.view_library)})


@router.get("/settings", response_model=ResponseBase[SystemSettingOut])
def get_permission_settings(db: DbSession, user: CurrentUser):
    return ResponseBase(data=SystemSettingOut(**PermissionService.get_settings(db)))


@router.put("/settings", response_model=ResponseBase[SystemSettingOut])
def update_permission_settings(
    db: DbSession,
    reviewer: AdminUser,
    block_upper_role_tasks: bool = Query(...),
):
    try:
        settings = PermissionService.update_settings(db, reviewer, block_upper_role_tasks)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    return ResponseBase(data=SystemSettingOut(**settings), message="设置已更新")

File: Information_Aggregation/frontend/src/api/permissions.ts
import request, { type ApiResponse, type PageResult } from './request'

export interface AccessRequest {
  id: number
  user_id: number
  request_type: string
  request_type_label?: string
  status: string
  reason: string | null
  reviewer_id: number | null
  review_note: string | null
  created_at: string
  reviewed_at: string | null
  username: string | null
  nickname: string | null
  reviewer_username: string | null
  reviewer_nickname: string | null
}

export interface RequestTypeOption {
  value: string
  label: string
}

export interface AccessRequestStats {
  my_pending: number
  my_total: number
  pending_for_review: number
  can_review: boolean
}

export function getApplicableRequestTypes() {
  return request.get<any, ApiResponse<RequestTypeOption[]>>('/permissions/access-requests/types')
}

export function getAccessRequestStats() {
  return request.get<any, ApiResponse<AccessRequestStats>>('/permissions/access-requests/stats')
}

export function submitAccessRequest(request_type: string, reason?: string) {
  return request.post<any, ApiResponse<AccessRequest>>('/permissions/access-requests', {
    request_type,
    reason,
  })
}

export function getAccessRequests(params: {
  status?: string
  request_type?: string
  scope?: 'mine' | 'review'
  page?: number
  page_size?: number
}) {
  return request.get<any, ApiResponse<PageResult<AccessRequest>>>('/permissions/access-requests', {
    params,
  })
}

export function reviewAccessRequest(requestId: number, approve: boolean, review_note?: string) {
  return request.post<any, ApiResponse<AccessRequest>>(
    `/permissions/access-requests/${requestId}/review`,
    { approve, review_note }
  )
}

export function revokeLibraryAccess(userId: number) {
  return request.post<any, ApiResponse<{ user_id: number; view_library: boolean }>>(
    `/permissions/users/${userId}/revoke-library`
  )
}

export function getPermissionSettings() {
  return request.get<any, ApiResponse<{ block_upper_role_tasks: boolean }>>('/permissions/settings')
}

export function updatePermissionSettings(block_upper_role_tasks: boolean) {
  return request.put<any, ApiResponse<{ block_upper_role_tasks: boolean }>>(
    '/permissions/settings',
    null,
    { params: { block_upper_role_tasks } }
  )
}

export const PERMISSION_LABELS: Record<string, string> = {
  view_library: '查阅达人库',
  view_all_meetings: '查阅全部会议',
  view_root_meetings: '查阅超管会议',
  view_all_root_meetings: '查阅全部超管会议',
  download_meetings: '会议导出/下载',
  approve_meeting_download: '审批会议下载',
}

export const REQUEST_TYPE_LABELS: Record<string, string> = {
  view_library: '查阅达人库',
  view_all_meetings: '查阅全部会议',
  download_meetings: '会议导出/下载',
  view_root_meetings: '查阅超管会议（限3天）',
  promote_admin: '升级为管理员',
}
